File: frontend/apps/ios/Pantopus/Features/ContentDetail/ContentDetailContent.swift
//
//  ContentDetailContent.swift
//  Pantopus
//
//  Render-only models for the T2.6 Transactional Detail shell.
//  `ContentDetailContent` is a flat description the shell consumes —
//  the per-entity view-models (gig / listing / invoice) project their
//  backend payload into one of these. Modules are an open enum so
//  the backend's jsonb_modules[] can extend the middle section
//  without touching the shell.
//

import Foundation

/// Which entity is being shown. Drives a few subtle treatments — the
/// listing variant gets a full-bleed cover image, the invoice variant
/// gets a full-width pay dock without a secondary button.
public enum ContentDetailKind: String, Sendable, Hashable {
    case gig
    case listing
    case invoice
}

/// Cover image for the listing variant (gradient placeholder + glyph
/// until the first photo loads). Optional — gigs and invoices skip
/// the cover slot.
public struct ContentDetailCover: Sendable, Hashable {
    public let imageUrl: URL?
    public let gradient: ListingGradient
    public let placeholderIcon: PantopusIcon
    public let pageCount: Int
    public let activePage: Int
    /// Sold treatment: desaturates the hero + stamps a tilted SOLD badge.
    public let sold: Bool
    /// Decorative glass action chips overlaid top-right (share / bookmark).
    /// Render-only for now — live wiring lands with the share + save flows.
    public let glassActions: [PantopusIcon]

    public init(
        imageUrl: URL?,
        gradient: ListingGradient,
        placeholderIcon: PantopusIcon,
        pageCount: Int = 1,
        activePage: Int = 0,
        sold: Bool = false,
        glassActions: [PantopusIcon] = []
    ) {
        self.imageUrl = imageUrl
        self.gradient = gradient
        self.placeholderIcon = placeholderIcon
        self.pageCount = pageCount
        self.activePage = activePage
        self.sold = sold
        self.glassActions = glassActions
    }
}

/// Top-of-content status pill — "Open · 4 bids", "Due in 3 days", etc.
public struct ContentDetailPill: Sendable, Hashable, Identifiable {
    public enum Tone: Sendable, Hashable { case info, success, warning, business, neutral, error }

    public let id: String
    public let label: String
    public let icon: PantopusIcon?
    public let tone: Tone

    public init(id: String = UUID().uuidString, label: String, icon: PantopusIcon? = nil, tone: Tone = .info) {
        self.id = id
        self.label = label
        self.icon = icon
        self.tone = tone
    }
}

/// Hero block under the status pill. `subtitle` is the category chip +
/// meta line; `priceLine` is the big number; `priceCaption` is the
/// faded suffix ("budget", "due", etc).
public struct ContentDetailHero: Sendable, Hashable {
    /// Drives the colour of the big price number. `auto` resolves to
    /// `primary600` for listings and `appText` elsewhere; `success`
    /// recolours it green for the invoice paid state.
    public enum PriceTone: Sendable, Hashable { case auto, success }

    public let title: String
    public let categoryChip: ContentDetailCategoryChip?
    public let meta: String?
    public let monoId: String?
    public let priceLine: String?
    public let priceCaption: String?
    public let priceTone: PriceTone
    /// Strikes through the price (listing sold — original asking price).
    public let priceStrikethrough: Bool
    /// Green sale tag rendered next to a struck-through price ("Sold for $385").
    public let saleTag: String?
    /// Right-aligned faded label trailing the price ("paid in full").
    public let priceTrailingLabel: String?
    /// Green check disc rendered after the price (invoice paid).
    public let priceCheckDisc: Bool
    /// Pill row rendered directly under the price (listing condition / pickup / distance).
    public let inlinePills: [ContentDetailPill]

    public init(
        title: String,
        categoryChip: ContentDetailCategoryChip? = nil,
        meta: String? = nil,
        monoId: String? = nil,
        priceLine: String? = nil,
        priceCaption: String? = nil,
        priceTone: PriceTone = .auto,
        priceStrikethrough: Bool = false,
        saleTag: String? = nil,
        priceTrailingLabel: String? = nil,
        priceCheckDisc: Bool = false,
        inlinePills: [ContentDetailPill] = []
    ) {
        self.title = title
        self.categoryChip = categoryChip
        self.meta = meta
        self.monoId = monoId
        self.priceLine = priceLine
        self.priceCaption = priceCaption
        self.priceTone = priceTone
        self.priceStrikethrough = priceStrikethrough
        self.saleTag = saleTag
        self.priceTrailingLabel = priceTrailingLabel
        self.priceCheckDisc = priceCheckDisc
        self.inlinePills = inlinePills
    }
}

/// Small category chip rendered next to the hero subtitle. Re-uses the
/// Gigs category palette (T2.3) so the vocabulary stays consistent.
public struct ContentDetailCategoryChip: Sendable, Hashable {
    public let label: String
    public let category: GigsCategory

    public init(label: String, category: GigsCategory) {
        self.label = label
        self.category = category
    }
}

/// One cell in the 3-up stat strip (gigs). Skipping the strip is fine
/// — empty array hides it.
public struct ContentDetailStat: Sendable, Hashable, Identifiable {
    public let id: String
    public let top: String
    public let bottom: String

    public init(id: String = UUID().uuidString, top: String, bottom: String) {
        self.id = id
        self.top = top
        self.bottom = bottom
    }
}

/// Counterparty card (avatar + name + identity + rating + message
/// button). Used by the listing variant for the seller; gigs surface
/// the bidder list as a module rather than a single counterparty.
public struct ContentDetailCounterparty: Sendable, Hashable {
    public let displayName: String
    public let initials: String
    public let avatarUrl: URL?
    public let identityKind: String? // "personal" | "business" | nil
    public let verified: Bool
    public let rating: Double?
    public let trailing: String?
    public let showsMessageButton: Bool

    public init(
        displayName: String,
        initials: String,
        avatarUrl: URL? = nil,
        identityKind: String?,
        verified: Bool,
        rating: Double?,
        trailing: String?,
        showsMessageButton: Bool = true
    ) {
        self.displayName = displayName
        self.initials = initials
        self.avatarUrl = avatarUrl
        self.identityKind = identityKind
        self.verified = verified
        self.rating = rating
        self.trailing = trailing
        self.showsMessageButton = showsMessageButton
    }
}

/// Trust capsule rendered in a wrap below the hero. Tone matches the
/// status-pill enum.
public typealias ContentDetailTrustCapsule = ContentDetailPill

/// Sticky dock button. `enabled == false` renders the disabled treatment
/// (sunken fill, muted text, no shadow) — used for the awarded gig's
/// "Bidding closed" lock CTA.
public struct ContentDetailDockButton: Sendable, Hashable {
    public let label: String
    public let icon: PantopusIcon?
    public let enabled: Bool

    public init(label: String, icon: PantopusIcon? = nil, enabled: Bool = true) {
        self.label = label
        self.icon = icon
        self.enabled = enabled
    }
}

/// Sticky dock CTA. `secondary == nil` → primary spans full width.
public struct ContentDetailDock: Sendable, Hashable {
    public let secondary: ContentDetailDockButton?
    public let primary: ContentDetailDockButton

    public init(secondary: ContentDetailDockButton? = nil, primary: ContentDetailDockButton) {
        self.secondary = secondary
        self.primary = primary
    }
}

// MARK: - Top-level content

/// Render-only content the shell consumes. The per-entity view-models
/// produce one of these from their backend payload.
public struct ContentDetailContent: Sendable, Hashable {
    public let kind: ContentDetailKind
    public let cover: ContentDetailCover?
    public let statusPill: ContentDetailPill?
    public let hero: ContentDetailHero
    public let statStrip: [ContentDetailStat]
    public let counterparty: ContentDetailCounterparty?
    public let modules: [ContentDetailModule]
    public let trustCapsules: [ContentDetailTrustCapsule]
    public let dock: ContentDetailDock

    public init(
        kind: ContentDetailKind,
        cover: ContentDetailCover? = nil,
        statusPill: ContentDetailPill? = nil,
        hero: ContentDetailHero,
        statStrip: [ContentDetailStat] = [],
        counterparty: ContentDetailCounterparty? = nil,
        modules: [ContentDetailModule] = [],
        trustCapsules: [ContentDetailTrustCapsule] = [],
        dock: ContentDetailDock
    ) {
        self.kind = kind
        self.cover = cover
        self.statusPill = statusPill
        self.hero = hero
        self.statStrip = statStrip
        self.counterparty = counterparty
        self.modules = modules
        self.trustCapsules = trustCapsules
        self.dock = dock
    }
}

/// Render state for the shell. View-models populate one of these.
public enum ContentDetailState: Sendable {
    case loading
    case loaded(ContentDetailContent)
    case error(message: String)
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/contentdetail/ContentDetailContent.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.ui.screens.contentdetail

import androidx.compose.runtime.Immutable
import app.pantopus.android.ui.screens.gigs.GigsCategory
import app.pantopus.android.ui.screens.marketplace.ListingGradient
import app.pantopus.android.ui.theme.PantopusIcon

/** Which entity is being shown. */
enum class ContentDetailKind { Gig, Listing, Invoice }

/** Cover image for the listing variant. */
@Immutable
data class ContentDetailCover(
    val imageUrl: String?,
    val gradient: ListingGradient,
    val placeholderIcon: PantopusIcon,
    val pageCount: Int = 1,
    val activePage: Int = 0,
    /** Sold treatment: desaturates the hero + stamps a tilted SOLD badge. */
    val sold: Boolean = false,
    /** Decorative glass action chips overlaid top-right (share / bookmark). */
    val glassActions: List<PantopusIcon> = emptyList(),
)

/** Status / trust pill. */
@Immutable
data class ContentDetailPill(
    val id: String,
    val label: String,
    val icon: PantopusIcon? = null,
    val tone: Tone = Tone.Info,
) {
    enum class Tone { Info, Success, Warning, Business, Neutral, Error }
}

@Immutable
data class ContentDetailCategoryChip(
    val label: String,
    val category: GigsCategory,
)

@Immutable
data class ContentDetailHero(
    val title: String,
    val categoryChip: ContentDetailCategoryChip? = null,
    val meta: String? = null,
    val monoId: String? = null,
    val priceLine: String? = null,
    val priceCaption: String? = null,
    /** `Auto` → primary600 for listings / appText elsewhere; `Success` → green. */
    val priceTone: PriceTone = PriceTone.Auto,
    /** Strikes through the price (listing sold). */
    val priceStrikethrough: Boolean = false,
    /** Green sale tag next to a struck price ("Sold for $385"). */
    val saleTag: String? = null,
    /** Right-aligned faded label trailing the price ("paid in full"). */
    val priceTrailingLabel: String? = null,
    /** Green check disc after the price (invoice paid). */
    val priceCheckDisc: Boolean = false,
    /** Pill row under the price (listing condition / pickup / distance). */
    val inlinePills: List<ContentDetailPill> = emptyList(),
) {
    enum class PriceTone { Auto, Success }
}

@Immutable
data class ContentDetailStat(
    val top: String,
    val bottom: String,
)

@Immutable
data class ContentDetailCounterparty(
    val displayName: String,
    val initials: String,
    val avatarUrl: String? = null,
    val identityKind: String?,
    val verified: Boolean,
    val rating: Double?,
    val trailing: String?,
    val showsMessageButton: Boolean = true,
)

@Immutable
data class ContentDetailDockButton(
    val label: String,
    val icon: PantopusIcon? = null,
    /** `false` renders the disabled treatment (sunken fill, muted text). */
    val enabled: Boolean = true,
)

@Immutable
data class ContentDetailDock(
    val secondary: ContentDetailDockButton? = null,
    val primary: ContentDetailDockButton,
)

// MARK: - Modules

sealed interface ContentDetailModule {
    val id: String

    @Immutable
    data class Description(
        override val id: String,
        val title: String,
        val icon: PantopusIcon? = null,
        val body: String,
    ) : ContentDetailModule

    @Immutable
    data class DetailRow(
        override val id: String,
        val title: String,
        val sectionIcon: PantopusIcon? = null,
        val rowIcon: PantopusIcon,
        val label: String,
        val trailing: String? = null,
    ) : ContentDetailModule

    @Immutable
    data class CaptionedText(
        override val id: String,
        val title: String,
        val icon: PantopusIcon? = null,
        val label: String,
    ) : ContentDetailModule

    @Immutable
    data class PhotoStrip(
        override val id: String,
        val title: String,
        val icon: PantopusIcon? = PantopusIcon.File,
        val countLabel: String? = null,
        val tiles: List<ContentDetailPhotoTile>,
    ) : ContentDetailModule

    /** Pickup → drop-off two-stop card (Magic Task V2). */
    @Immutable
    data class TwoStop(
        override val id: String,
        val title: String,
        val icon: PantopusIcon? = PantopusIcon.MapPin,
        val stops: List<Stop>,
    ) : ContentDetailModule {
        enum class StopTone { Primary, Success }

        @Immutable
        data class Stop(
            val letter: String,
            val tone: StopTone,
            val address: String,
            val distance: String?,
        )
    }

    /** Inline wrap of trust/status capsules placed mid-flow. */
    @Immutable
    data class CapsuleRow(
        override val id: String,
        val capsules: List<ContentDetailPill>,
    ) : ContentDetailModule

    /** Key/value detail grid (listing "Details"). */
    @Immutable
    data class DetailsGrid(
        override val id: String,
        val title: String,
        val icon: PantopusIcon? = PantopusIcon.AlertCircle,
        val rows: List<Row>,
    ) : ContentDetailModule {
        @Immutable
        data class Row(val key: String, val value: String)
    }

    /**
     * Flexible callout card — awarded banner, Pantopus Pay receipt
     * capsule, "Alert me when similar appears" row, no-bids empty capsule.
     */
    @Immutable
    data class Callout(
        override val id: String,
        val style: Style = Style.Banner,
        val tone: Tone = Tone.Success,
        val icon: PantopusIcon,
        val iconTone: IconTone = IconTone.Success,
        val title: String,
        val subtitle: String? = null,
        val subtitleMono: Boolean = false,
        val trailingActionLabel: String? = null,
        val footerPill: String? = null,
    ) : ContentDetailModule {
        enum class Style { Banner, Empty }

        enum class Tone { Success, Neutral, Dashed }

        enum class IconTone { Success, SuccessOutline, Primary }
    }

    @Immutable
    data class SimilarStrip(
        override val id: String,
        val title: String,
        val sub: String? = null,
        val items: List<ContentDetailSimilarItem>,
    ) : ContentDetailModule

    @Immutable
    data class Bids(
        override val id: String,
        val title: String,
        val sub: String? = null,
        val bids: List<ContentDetailBidRow>,
    ) : ContentDetailModule

    @Immutable
    data class FromTo(
        override val id: String,
        val from: ContentDetailParty,
        val to: ContentDetailParty,
    ) : ContentDetailModule

    @Immutable
    data class LineItems(
        override val id: String,
        val title: String,
        val icon: PantopusIcon? = PantopusIcon.File,
        val rows: List<ContentDetailLineItem>,
        val fees: List<ContentDetailSummaryRow> = emptyList(),
        val totalLabel: String? = null,
        val totalValue: String? = null,
        val totalTone: TotalTone = TotalTone.Primary,
    ) : ContentDetailModule {
        enum class TotalTone { Primary, Success }
    }

    @Immutable
    data class Summary(
        override val id: String,
        val rows: List<ContentDetailSummaryRow>,
        val totalLabel: String,
        val totalValue: String,
        val totalTone: TotalTone = TotalTone.Primary,
    ) : ContentDetailModule {
        enum class TotalTone { Primary, Success }
    }

    /** Interactive mini map — tap to open full-screen explorer. */
    @Immutable
    data class LocationMap(
        override val id: String = "location_map",
        val latitude: Double,
        val longitude: Double,
        val isApproximate: Boolean,
        val footnote: String,
        val category: GigsCategory,
    ) : ContentDetailModule
}

@Immutable
data class ContentDetailPhotoTile(
    val id: String,
    val gradient: ListingGradient,
    val icon: PantopusIcon,
    /**
     * Real attachment image. When set the tile renders the photo and falls
     * back to the gradient + glyph while loading / on failure.
     */
    val imageUrl: String? = null,
)

@Immutable
data class ContentDetailSimilarItem(
    val id: String,
    val title: String,
    val price: String,
    val gradient: ListingGradient,
)

@Immutable
data class ContentDetailBidRow(
    val id: String,
    val initials: String,
    val displayName: String,
    val ratingLine: String,
    val amount: String,
    val verified: Boolean,
    /** Optional tag pill ("fastest reply" / "has van"). */
    val tag: String? = null,
    /** Winning bid (awarded) — green tint + Winner pill + green amount. */
    val won: Boolean = false,
    /** Losing bid (awarded) — 55% opacity + struck-through amount. */
    val dimmed: Boolean = false,
)

@Immutable
data class ContentDetailParty(
    val label: String,
    val name: String,
    val sub: String,
    val accent: Accent,
) {
    enum class Accent { Business, Personal, Neutral }
}

@Immutable
data class ContentDetailLineItem(
    val id: String,
    val item: String,
    val qty: String,
    val unit: String,
    val total: String,
)

@Immutable
data class ContentDetailSummaryRow(
    val id: String,
    val label: String,
    val value: String,
)

@Immutable
data class ContentDetailContent(
    val kind: ContentDetailKind,
    val cover: ContentDetailCover? = null,
    val statusPill: ContentDetailPill? = null,
    val hero: ContentDetailHero,
    val statStrip: List<ContentDetailStat> = emptyList(),
    val counterparty: ContentDetailCounterparty? = null,
    val modules: List<ContentDetailModule> = emptyList(),
    val trustCapsules: List<ContentDetailPill> = emptyList(),
    val dock: ContentDetailDock,
)

sealed interface ContentDetailUiState {
    data object Loading : ContentDetailUiState

    data class Loaded(val content: ContentDetailContent) : ContentDetailUiState

    data class Error(val message: String) : ContentDetailUiState
}
